tools/app.hpp

```diff
@@ -128,34 +129,43 @@ namespace arg {
         {
             auto* wand = app->add_option("-w,--wand", m_wand_data_path, "WAND data filename");
             auto* scorer =
-                app->add_option("-s,--scorer", m_scorer, "Query processing algorithm")->needs(wand);
+                app->add_option("-s,--scorer", m_params.name, "Query processing algorithm")->needs(wand);
             app->add_flag("--quantize", m_quantize, "Quantizes the scores")->needs(scorer);
+            app->add_option("--bm25-k1", m_params.bm25_k1, "BM25 k1 parameter.");
+            app->add_option("--bm25-b", m_params.bm25_b, "BM25 b parameter.");
+            app->add_option("--pl2-c", m_params.pl2_c, "PL2 c parameter.");
+            app->add_option("--qld-mu", m_params.qld_mu, "QLD mu parameter.");
         }
 
-        [[nodiscard]] auto scorer() const -> std::optional<std::string> const& { return m_scorer; }
+        [[nodiscard]] auto scorer_params() const { return m_params; }
+
         [[nodiscard]] auto wand_data_path() const -> std::optional<std::string> const&
         {
             return m_wand_data_path;
         }
+
         [[nodiscard]] auto quantize() const { return m_quantize; }
 
       private:
-        std::optional<std::string> m_scorer;
+        ScorerParams m_params;
         std::optional<std::string> m_wand_data_path;
         bool m_quantize = false;
     };
 
     struct Scorer {
         explicit Scorer(CLI::App* app)
         {
-            auto* opt =
-                app->add_option("-s,--scorer", m_scorer, "Query processing algorithm")->required();
+            app->add_option("-s,--scorer", m_params.name, "Query processing algorithm")->required();
+            app->add_option("--bm25-k1", m_params.bm25_k1, "BM25 k1 parameter.");
+            app->add_option("--bm25-b", m_params.bm25_b, "BM25 b parameter.");
+            app->add_option("--pl2-c", m_params.pl2_c, "PL2 c parameter.");
+            app->add_option("--qld-mu", m_params.qld_mu, "QLD mu parameter.");
```

Comment:
Since these are the same as the ones below, maybe we can extract it to a template function:
```cpp
template<typename T>
void add_scorer_options(CLI::App* app, T& args) {
    app->add_option("-s,--scorer", args.m_params.name, "Scorer function")->required();
    ...
}
```
and call it with `add_scorer_options(app, *this)`.

Having the same thing in two places is bug-prone. Case in point:
```
app->add_option("-s,--scorer", m_params.name, "Query processing algorithm")->required();
                                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^

```
Happens in one place but not the other.

Reply:
Thanks, good idea. I am looking at this now and I noticed that if I make this a free function, I can't access the `ScorerParams` object since it's a `private` member inside each struct.

I've opted to make the free function a friend of each struct that requires it. Not entirely sure if this is a good solution,  let me know what you think.